sympy/physics/control/tests/test_lti.py

```diff
@@ -0,0 +1,165 @@
+
+    H1 = TransferFunction(
+        Matrix([7 / (p**9 + 9*p)])
+    )
+    H2 = TransferFunction(
+        Matrix([5 / (s + 2)])
+    )
+    H = TransferFunction(
+        Matrix([7 / (p**9 + 9*p) + 5 / (s + 2)])
+    )
+    assert together(H1.parallel(H2).G) == together(H.G)
+
+
+def test_TransferFunction_negate():
+    G1 = Matrix([s / (s + 1)])
+    tf1 = TransferFunction(G1)
+    assert tf1.neg() == TransferFunction(
+        Matrix([-s / (s + 1)])
+    )
+
+    G2 = Matrix([[(4*p - 10) / (2*p + 1), 3 / (p + 2)],
+            [1 / (2*p**2 + 5*p + 2), (p + 1) / (p**2 + 4*p + 4)]])
+    tf2 = TransferFunction(G2)
+    assert tf2.neg() == TransferFunction(
+        Matrix([[(10 - 4*p) / (2*p + 1),  -3 / (p + 2)],
+            [-1 / (2*p**2 + 5*p + 2), (-p - 1) / (p**2 + 4*p + 4)]])
+    )
+
+
+def test_TransferFunction_is_proper():
+    G1 = Matrix([2*p / (2 + p**2 - p), p / (1 + p**3)])
+    assert TransferFunction(G1).is_proper
+
+    G2 = Matrix([2*s**6/(2 + s**2 - s), (s**4)/(1 + s**2)])
+    assert not TransferFunction(G2).is_proper
+
+
+def test_TransferFunction_is_strictly_proper():
+    G1 = Matrix([2*s**6 / (2 + s**2 - s), (s**4) / (1 + s**2)])
+    assert not TransferFunction(G1).is_strictly_proper
+
+    G2 = Matrix([[(4*p - 10) / (2*p**3 + 1), 3 / (p + 2)],
+            [1 / (2*p**2 + 5*p + 2), (p + 1) / (p**2 + 4*p + 4)]])
+    assert TransferFunction(G2).is_strictly_proper
+
+
+def test_SISOTransferFunction_create():
+    G1 = SISOTransferFunction(s + 1, s**2 + s + 1)
+    assert G1.num == (s + 1)
+    assert G1.den == (s**2 + s + 1)
+
+    G2 = SISOTransferFunction(p + 3, p**2 - 9)
+    assert G2.num == (p + 3)
+    assert G2.den == (p**2 - 9)
+
+    G3 = SISOTransferFunction(s + 4, s - 5)
+    assert G3.num == (s + 4)
+    assert G3.den == (s - 5)
+
+
+def test_SISOTransferFunction_series():
+    G1 = SISOTransferFunction(s + 3, -s**3 + 9)
+    G2 = SISOTransferFunction(s + 1, s - 5)
+    expect = SISOTransferFunction(-s**2 - 4*s - 3, s**4 - 5*s**3 -9*s + 45)
+    assert G1.series(G2) == expect
+
+    G3 = SISOTransferFunction(p, p**4 - 6)
+    G4 = SISOTransferFunction(p + 4, p - 5)
+    expect = SISOTransferFunction(p**2 + 4*p, p**5 - 5*p**4 - 6*p + 30)
+    assert G3.series(G4) == expect
+
+
+def test_SISOTransferFunction_negate():
+    tf1 = SISOTransferFunction(s + 3, s**2 - s**3 + 9)
+    assert tf1.neg() == SISOTransferFunction(-s - 3, s**2 - s**3 + 9)
+
+    tf2 = SISOTransferFunction(-3*p + 3, 1 - p)
+    assert tf2.neg() == SISOTransferFunction(3*p - 3, 1 - p)
+
+
+def test_SISOTransferFunction_is_proper():
+    G1 = SISOTransferFunction(s**4 + s**3 - 2*s, s**4 - 1)
+    assert G1.is_proper
+
+    G2 = SISOTransferFunction(p**7, p - 4)
+    assert not G2.is_proper
+
+
+def test_SISOTransferFunction_is_strictly_proper():
+    tf1 = SISOTransferFunction(s**3 - 2, s**4 + 5*s + 6)
+    assert tf1.is_strictly_proper
+
+    tf2 = SISOTransferFunction(p + 3, p - 4)
+    assert not tf2.is_strictly_proper
```

Comment:
This is a pretty minimal set of unit tests. I don't think it is sufficient to fully understand what the API needs to be and what functionality is needed.

Here are some suggestions of what the unit tests for a SISO transfer function could include:

- Instantiation of immutable transfer functions
  - Using different polynomial variables
  - Inputs that are not polynomials and what errors this raises
  - Purely symbolic poylynomials
  - Show that pole, zero cancelation does not occur
- Method for simplifying the transfer function, or how it works with simplify(), this causes pole zero cancelations
- Whether any of the operators should work: `__add__()`, `__mul__()`, `__div__()`, etc., what are operations with other transferfunctions, other expressions, other variables, etc.
- What does the srerp, repr, and str() versions of this class produce
- Evaluating the transfer function numerically given a purely symbolic TFs
- Converting to zero, pole, gain, form
- For low order symbolic transfer functions, have a way to covert to a time domain form, i.e. solve the ODE using the ODE module(s)
- Method to calculate the zeros and poles: symbolically for low order and numerically for high order
- How do standard sympy functions work with TFs: subs(), xreplace(), expand(), factor(), evalf(), etc.?
- Is the Polynomial class used internally in any way, if so you need some tests to show what you want to do with this.
- Method to apply different symbolic input functions in the Laplace domain to the transfer function: step, ramp, impulse, etc.
- Methods for block diagram algebra
- etc, etc, etc.

Until you express a large set of the desired functionality for this class as a set of comprehensive unit tests, we can understand what the proposal is. Once you have unit tests for all or most of the above, then we can review and give you feedback on the class API and functionality design.

Reply:
Thanks, I will study these extra topics in more detail and first work on a comprehensive set of unit tests. Cause I thought only the most basic functionality was included in my proposal, so I didn't prepare for that before. But I can work on this now :)

Reply:
You should think of your GSoC proposal as showing us what topic you'd like to contribute to SymPy and that you can succeed at contributing that topic. The GSoC proposals are rarely implemented exactly as described and are not vetted in a way that is the same as a review of what code is merged into SymPy. Your GSoC proposal showed that you have a good starting point. About half of the things I've listed above are the minimum things you need in place to get a new class merged into SymPy that follows SymPy design patterns. The other half are the methods that you'd expect a useful transfer function class to have.

Reply:
I've reorganized the list here into the minimal functionality and other functionality. It is important to think about all of this functionality so you don't design the core class in a way that prohibits or makes it difficult to implement the extended set of features.

# Minimal

- Instantiation of immutable transfer functions
  - Using different polynomial variables
  - Inputs that are not polynomials and what errors this raises
  - Purely symbolic poylynomials
  - Show that pole, zero cancelation does not occur
- Method for simplifying the transfer function, or how it works with simplify(), this causes pole zero cancelations
- Whether any of the operators should work: `__add__()`, `__mul__()`, `__div__()`, etc., what are operations with other transferfunctions, other expressions, other variables, etc.
- What does the srerp, repr, and str() versions of this class produce
- Evaluating the transfer function numerically given a purely symbolic TFs
- How do standard sympy functions work with TFs: subs(), xreplace(), expand(), factor(), evalf(), etc.?
- Is the Polynomial class used internally in any way, if so you need some tests to show what you want to do with this.
- Printing: latex, pretty printer, code printers, etc.

# Other
- Converting to zero, pole, gain, form
- For low order symbolic transfer functions, have a way to covert to a time domain form, i.e. solve the ODE using the ODE module(s)
- Method to calculate the zeros and poles: symbolically for low order and numerically for high order
- Method to apply different symbolic input functions in the Laplace domain to the transfer function: step, ramp, impulse, etc.
- Methods for block diagram algebra and feedback
- Methods for taking the limit as `s` goes to zero and infinity: zero frequency gains, initial values, steady state
- Plots: Zero-Pole plot, Bode, Root locus, Nichols, etc. 
- etc, etc, etc.

Reply:
Thanks for mentioning all the required functionality for this class. I'm writing examples to showcase all the methods and features I will add in this. Will be ready soon. I hope I can start the implementation after that. 

> You should think of your GSoC proposal as showing us what topic you'd like to contribute to SymPy and that you can succeed at contributing that topic. The GSoC proposals are rarely implemented exactly as described and are not vetted in a way that is the same as a review of what code is merged into SymPy. Your GSoC proposal showed that you have a good starting point.

Oh I see. Right now, I'm just a little worried that I might not be able to complete all the things mentioned in my proposal within three months.. because new things have been introduced. But I'm excited about learning new stuff, so I will give more time to it. Thanks, though.

Reply:
> Right now, I'm just a little worried that I might not be able to complete all the things mentioned in my proposal within three months.

I recommend not worrying too much about this. Focus on getting quality contributions that build out the control systems functionality in SymPy and putting in your weekly required time for GSoC. At the end you will have something to be proud of.

Reply:
Thanks. 

> Evaluating the transfer function numerically given a purely symbolic TFs

Can you mention more about this one.. like what kind of function should I design? Something like [this](https://www.cds.caltech.edu/~murray/wiki/How_do_I_evaluate_a_certain_transfer_function_at_desired_frequencies_numerically%3F) which python control has?

Reply:
Here is an example:

```python
>>> s = symbols('s')
>>> a0, a1, b0, b1, b2 = symbols('a0, a1, b0, b1, b2')
>>> p1 = a1*s + a0
>>> p2 = b2*s**2 + b1*s + b0
>>> tf = TransferFunction(p1, p2, var=s)
>>> tf.evalf(subs={a0: 1, a1: 2, b0: 3, b1: 4, b2: 5})
TransferFunction(2*s + 1, 5*s**2 + 4*s + 3)
```

Reply:
> Can you mention more about this one.. like what kind of function should I design? Something like this which python control has?

This is different functionality. I think this might work like:

```python
>>> s = symbols('s')
>>> num = 2*s + 1
>>> den = 3*s**2 + 2*s + 1
>>> tf = TransferFunction(num, den)
>>> import numpy as np
>>> omegas = np.logspace(0, 1000)
>>> eval_tf = lambdify(s, tf)
>>> magnitudes, phases = eval_tf(omegas)
```

Reply:
Okay! I think that in this PR, we can cover the minimal functionality (as you stated above) for `TransferFunction` and other functionality can be covered in the following PRs when I'll add `bode` and `root_locus` plots. for instance - Methods like finding zeros and poles would be essential for `root_locus` plot.

Reply:
@namannimmo10 I've looked through all the tests and made comments. You are also still missing a number of tests for the things in the "minimal" list above.

Reply:
Okay, I will add the remaining ones soon.